py/dml/breaking_changes.py

```diff
@@ -0,0 +1,495 @@
+
+    The most immediate effect of enabling this change
+    is that DMLC will report errors on statements like `assert 0;` and
+    `while (1) { ... }`, which need to change into `assert false;` and
+    `while (true) { ... }`, respectively. Other effects include:
+
+    * With legacy integer semantics, integers of non-standard sizes
+      are represented as a native C
+      type, e.g. `uint5` is represented as `uint8`, allowing it to
+      store numbers too large to fit in 5 bits. With modern integer
+      semantics, arithmetic is done on 64-bit integers and bits are
+      truncated if casting or storing in a smaller type.
+
+      Old code sometimes relies on this feature by comparing variables
+      of type `int1` to the value `1`. In DML 1.4, the only values of
+      type `int1` are `0` and `-1`, so such code should be rewritten
+      to use the `uint1` type. It can be a good idea to grep for
+      `[^a-z_]int1[^0-9]` and review if `uint1` is a better choice.
+
+    * With legacy semantics, some operations that have undefined behaviour
+      in C effectively counts as undefined behaviour in DML as well,
+      whereas modern semantics have well-defined semantics for
+      all operations. For instance, negative
+      shift or division by zero is handled with an unconditional critical
+      error, and too large shift operands or signed shift overflow is handled
+      by truncation.
+
+    * With legacy integer semantics, comparison operators `<`, `<=`, `==`,
+      `>=`, `>` inherit C semantics, whereas with modern integer semantics
+      they are compared mathematically as integers. This sometimes makes a
+      difference when comparing signed and unsigned numbers; in particular, `-1 !=
+      0xffffffffffffffff` consistently with modern semantics, whereas with
+      legacy semantics, they are consiered different only if
+      both are constant.
+    '''
+    short = "Apply modern integer semantics in DML 1.2"
+    required_after = api_6
+
+
+@change
+class forbid_experimental_vect(BreakingChange):
+    '''<a id="forbid_experimental_vect"/>
+
+    Up to Simics 7, the `vect` syntax was permitted without enabling the
+    [`simics_util_vect` provisional feature](provisional-auto.html#simics_util_vect),
+    issuing only a warning in DML 1.4.
+    When this change is enabled, the `vect` syntax is forbidden unless the
+    provisional feature is explicitly enabled.
+
+    where the [`simics_util_vect` provisional feature](provisional-auto.html#simics_util_vect)
+    is not enabled.
+
+    '''
+    short = "Disallow vect syntax without 'provisional simics_util_vect'"
+    required_after = api_7
+
+
+@change
+class forbid_warning_statement(BreakingChange):
+    '''
+    Up to Simics 7, the `_warning` statement was allowed, though rarely useful.
+    When this change is enabled, `_warning` statements are disallowed.
+    '''
+    short = "Disallow the `_warning` statement"
+    required_after = api_7
+
+
+
+compat_features = {
```

Comment:
once we eliminate uses of dml.globals.enabled_compat, this should be moved to its only use in dmlc.py